smithy-utils/src/main/java/software/amazon/smithy/utils/AbstractCodeWriter.java

```diff
@@ -426,6 +426,142 @@
  *     handleA,
  *     handleB);
  * }</pre>
+ *
+ * <h3>Template conditionals and loops</h3>
+ *
+ * <p>AbstractCodeWriter is a lightweight template engine that supports conditional
+ * blocks and loops.
+ *
+ * <h4>Conditional blocks</h4>
+ *
+ * <p>Conditional blocks are defined using the following syntax:
+ *
+ * <pre>{@code
+ * ${?foo}
+ * Foo is set: ${foo:L}
+ * ${/foo}
+ * }</pre>
+ *
+ * <p>Assuming {@code foo} is truthy and set to "hi", then the above template
+ * outputs: "Foo is set: hi"
+ *
+ * <p>In the above example, "?" indicates that the expression is a conditional block
+ * to check if the named parameter "foo" is truthy. If it is, then the contents of the
+ * block up to the matching closing block, {@code ${/foo}}, are evaluated. If the
+ * condition is not satisfied, then contents of the block are skipped.
+ *
+ * <p>You can check if a named property is falsey using "^":
+ *
+ * <pre>{@code
+ * ${^foo}
+ * Foo is not set
+ * ${/foo}
+ * }</pre>
+ *
+ * <p>Assuming {@code foo} is set to "hi", then the above template outputs nothing.
+ * If {@code foo} is falsey, then the above template output "Foo is not set".
+ *
+ * <h4>Truthy and falsey</h4>
+ *
+ * <p>The following values are considered falsey:
+ *
+ * <ul>
+ *     <li>values that are not found</li>
+ *     <li>null values</li>
+ *     <li>false</li>
+ *     <li>empty {@link String}</li>
+ *     <li>empty {@link Iterable}</li>
+ *     <li>empty {@link Map}</li>
+ *     <li>empty {@link Optional}</li>
+ * </ul>
+ *
+ * <p>Values that are not falsey are considered truthy.
+ *
+ * <h4>Loops</h4>
+ *
+ * <p>Loops can be created to repeat a section of a template for each key value pair
+ * stored in a named property. Loops are created using "#".
+ *
+ * <p>The following template with a "foo" value of ["a", "b", "c"]:
```

Comment:
You say key-value pair but then immediately follow up with a list. Either this first example should use a map, or the explanation that iterables use the index as a pseudo-key needs to come before this example.